### PR TITLE
User CRS implementation, authorization and resource access verification

### PR DESCRIPTION
**Pull request by dawba**
Follow up to changes proposed by @dimmonn and @Konrad1w.

Based on the authentication flow I have updated the controllers for User/Advertisement/AdvertisementImage/Category.
1. For User, Advertisement, AdvertisementImage
I concluded that to continue with DELETE/PUT mappings, we need to check if the user has the access to the resource as the owner, or is a super-owner of the resource (Admin). 

2. For Category
I concluded that only super-owner should have access to this resource in DELETE/PUT mappings.

To achieve requirements in 1 and 2, we end up with `ResourceAccessAuthorizationService` that is wired to our controllers.
This service provides us with two methods for access authorization - 
- `authorizeUserAccess` for user authorization from just the resourceId and resourceType. This allows us to be pretty general with the method usage, as it is reusable in all controllers. Based on resourceType we call subsequent methods to get to the User, as our database has links between User<-Ad<-AdImage
- `authorizeUserAccessFromRequestBodyOrThrow` for user authorization from the request body as we dont always pass down the resourceId. From what I've read, the `PreAuthorize` method does run BEFORE the request is processed (as in called based on the api path), so naturally this needs to be called when processing the actual request. We call it at the start of the request to determine if the user has the access, otherwise we throw the `AcessDenied` exception and move on:)

### CODE REVIEW

File: src/main/java/org/marketplace/configuration/DataLoader.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    @Autowired
+    private UserManagementRepository userRepository;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+    @Override
+    public void onApplicationEvent(ApplicationReadyEvent event) {
+        if (userRepository.count() == 0) {
+            User user = new User();
+            user.setEmail("user1@test.com");
```
Comment (dimmonn) [inline src/main/java/org/marketplace/configuration/DataLoader.java:25]: i could imagine that we could hide default user and pass behind linux env properties and autowire them into runtime. just to not stop the work shall i fill out an issue?

File: src/main/java/org/marketplace/controllers/AdvertisementImageManagementController.java [hunk 2]
```diff
@@ -54,6 +61,7 @@ public Response<AdvertisementImage> requestUpdateAdvertisementImage(@RequestBody
      * Delete an advertisement image
      * @param id id of the advertisement image to be deleted
      */
+    @PreAuthorize("@resourceAccessAuthorizationService.authorizeUserAccess('ResourceType.AdvertisementImage', #id).equals(T(org.marketplace.enums.AccessStatus).ACCESS_GRANTED)")
```
Comment (dimmonn) [inline src/main/java/org/marketplace/controllers/AdvertisementImageManagementController.java:64]: maybe i didn't understand this correctly, why we don't want to do something like?

```
@PreAuthorize("hasAuthority('SOME_AUTHORITY')")
```

Comment (dawba) [reply src/main/java/org/marketplace/controllers/AdvertisementImageManagementController.java:64]: We discussed this issue during today's lecture.
As a followup you proposed some different kind of approach that links the user and his created resource for management access.

Could you explain it further with some ideas that we can read about?

File: src/main/java/org/marketplace/configuration/DataLoader.java (hunk 1 above)
Comment (dawba) [reply src/main/java/org/marketplace/configuration/DataLoader.java:25]: Good idea, I have filled out an issue #22 and will get to work with that.

File: src/main/java/org/marketplace/controllers/AdvertisementImageManagementController.java (hunk 2 above)
Comment (dimmonn) [reply src/main/java/org/marketplace/controllers/AdvertisementImageManagementController.java:64]: sure, it's called domain object security when we answer on 3 questions, who, where (method level), what (object level), ill try to come up with some example for you 👍

**Comment by dawba** [comment]
As the discussion was resolved, issue for env variables created, I am closing this pull request - I need to proceed with token expiration (to allow users to log out) and implement some base for search feature.

File: src/main/java/org/marketplace/controllers/AdvertisementImageManagementController.java (hunk 2 above)
Comment (dimmonn) [reply src/main/java/org/marketplace/controllers/AdvertisementImageManagementController.java:64]: i just created an example PR of an ACL, https://github.com/dawba/market-place/pull/23
here is some general info about this approach https://github.com/eugenp/tutorials/tree/master/spring-security-modules/spring-security-acl

**Pull request merged**